File: backend-spring-boot/src/test/java/no/vicx/backend/user/vm/ChangePasswordVmTest.java
package no.vicx.backend.user.vm;

import no.vicx.database.user.VicxUser;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.security.crypto.password.PasswordEncoder;

import static org.junit.jupiter.api.Assertions.assertSame;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class ChangePasswordVmTest {

    @Mock
    VicxUser vicxUser;

    @Mock
    PasswordEncoder passwordEncoder;

    @Test
    void applyPatch_givenValidRequest_expectUpdatedVicxUser() {
        when(passwordEncoder.encode(anyString())).thenReturn("~encoded-password~");

        var sut = new ChangePasswordVm("~current-password~", "~new-password~");

        var returnedUser = sut.applyPatch(vicxUser, passwordEncoder);

        assertSame(vicxUser, returnedUser);
        verify(vicxUser).setPassword("~encoded-password~");

        verify(passwordEncoder).encode("~new-password~");
    }
}

File: backend-spring-boot/src/test/kotlin/no/vicx/backend/user/vm/ChangePasswordVmTest.kt
package no.vicx.backend.user.vm

import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.types.shouldBeSameInstanceAs
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import no.vicx.database.user.VicxUser
import org.springframework.security.crypto.password.PasswordEncoder

class ChangePasswordVmTest : BehaviorSpec({
    val vicxUser: VicxUser = mockk(relaxed = true)
    val passwordEncoder: PasswordEncoder = mockk()

    Given("valid change password view model") {
        val sut = ChangePasswordVm("~current-password~", "~new-password~")

        val expectedPassword = "~encoded-password~"
        every { passwordEncoder.encode(any()) } returns expectedPassword

        When("applying patch to user") {
            val updatedUser = sut.applyPatch(vicxUser, passwordEncoder)

            Then("expect user to be updated with new password") {
                updatedUser shouldBeSameInstanceAs vicxUser

                verify { vicxUser.password = expectedPassword }
            }
        }
    }
})